Misleading error message on associated type bounds

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
trait MyTrait {
    type Assoc;
}

#[allow(dead_code)]
fn foo<S, T>(_s: S, _t: T)
where
    S: MyTrait, T: MyTrait<Assoc == S::Assoc>
{}
```

Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=adddc330eab816444616202cc576f88e

I expected to see this happen: The error message should indicate that the `==` should be a single `=`.

Instead, this happened: The error message gives no hint of the correct syntax.  It gives two errors, neither of which are helpful:

```
error: expected one of `,` or `>`, found `==`
 --> src/lib.rs:8:34
  |
8 |     S: MyTrait, T: MyTrait<Assoc == S::Assoc>
  |                                  ^^ expected one of `,` or `>`
  |
help: expressions must be enclosed in braces to be used as const generic arguments
  |
8 |     S: MyTrait, T: MyTrait<{ Assoc == S::Assoc }>
  |                            ^                   ^
```

This error is unhelpful because it says "expected one of `,` or `>`", but `=` is also allowed.  Also, the suggestion to use curly braces is a red herring.

```
error[E0107]: this trait takes 0 const arguments but 1 const argument was supplied
 --> src/lib.rs:8:20
  |
8 |     S: MyTrait, T: MyTrait<Assoc == S::Assoc>
  |                    ^^^^^^^------------------- help: remove these generics
  |                    |
  |                    expected 0 const arguments
  |
note: trait defined here, with 0 const parameters
 --> src/lib.rs:1:7
  |
1 | trait MyTrait {
  |       ^^^^^^^
```

This error is unhelpful because I am not specifying const arguments; I'm specifying associated types.

### Meta

Rust version according to play.rust-lang.org: 1.53.0

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -1959,11 +1959,11 @@ impl<'a> Parser<'a> {
                 if token::EqEq == snapshot.token.kind {
                     err.span_suggestion(
                         snapshot.token.span,
-                        "replace `==` with `=`",
+                        "if you meant to use an associated type binding, replace `==` with `=`",
                         "=".to_string(),
                         Applicability::MaybeIncorrect,
                     );
-                    let value = self.mk_expr_err(expr.span);
+                    let value = self.mk_expr_err(start.to(expr.span));
                     err.emit();
                     return Ok(GenericArg::Const(AnonConst { id: ast::DUMMY_NODE_ID, value }));
                 } else if token::Comma == self.token.kind || self.token.kind.should_end_const_arg()

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -1955,7 +1955,19 @@ impl<'a> Parser<'a> {
         }
         match self.parse_expr_res(Restrictions::CONST_EXPR, None) {
             Ok(expr) => {
-                if token::Comma == self.token.kind || self.token.kind.should_end_const_arg() {
+                // Find a mistake like `MyTrait<Assoc == S::Assoc>`.
+                if token::EqEq == snapshot.token.kind {
+                    err.span_suggestion(
+                        snapshot.token.span,
+                        "replace `==` with `=`",
+                        "=".to_string(),
+                        Applicability::MaybeIncorrect,
+                    );
+                    let value = self.mk_expr_err(expr.span);
+                    err.emit();
+                    return Ok(GenericArg::Const(AnonConst { id: ast::DUMMY_NODE_ID, value }));
+                } else if token::Comma == self.token.kind || self.token.kind.should_end_const_arg()
+                {
                     // Avoid the following output by checking that we consumed a full const arg:
                     // help: expressions must be enclosed in braces to be used as const generic
                     //       arguments

--- a/src/test/ui/const-generics/issues/issue-87493.rs
+++ b/src/test/ui/const-generics/issues/issue-87493.rs
@@ -1,0 +1,14 @@
+pub trait MyTrait {
+    type Assoc;
+}
+
+pub fn foo<S, T>(_s: S, _t: T)
+where
+    S: MyTrait,
+    T: MyTrait<Assoc == S::Assoc>,
+    //~^ ERROR: expected one of `,` or `>`, found `==`
+    //~| ERROR: this trait takes 0 generic arguments but 1 generic argument was supplied
+{
+}
+
+fn main() {}

--- a/src/test/ui/const-generics/issues/issue-87493.stderr
+++ b/src/test/ui/const-generics/issues/issue-87493.stderr
@@ -2,10 +2,12 @@ error: expected one of `,` or `>`, found `==`
   --> $DIR/issue-87493.rs:8:22
    |
 LL |     T: MyTrait<Assoc == S::Assoc>,
-   |                      ^^
-   |                      |
-   |                      expected one of `,` or `>`
-   |                      help: replace `==` with `=`: `=`
+   |                      ^^ expected one of `,` or `>`
+   |
+help: if you meant to use an associated type binding, replace `==` with `=`
+   |
+LL |     T: MyTrait<Assoc = S::Assoc>,
+   |                      ~
 
 error[E0107]: this trait takes 0 generic arguments but 1 generic argument was supplied
   --> $DIR/issue-87493.rs:8:8

--- a/src/test/ui/const-generics/issues/issue-87493.stderr
+++ b/src/test/ui/const-generics/issues/issue-87493.stderr
@@ -1,0 +1,26 @@
+error: expected one of `,` or `>`, found `==`
+  --> $DIR/issue-87493.rs:8:22
+   |
+LL |     T: MyTrait<Assoc == S::Assoc>,
+   |                      ^^
+   |                      |
+   |                      expected one of `,` or `>`
+   |                      help: replace `==` with `=`: `=`
+
+error[E0107]: this trait takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/issue-87493.rs:8:8
+   |
+LL |     T: MyTrait<Assoc == S::Assoc>,
+   |        ^^^^^^^------------------- help: remove these generics
+   |        |
+   |        expected 0 generic arguments
+   |
+note: trait defined here, with 0 generic parameters
+  --> $DIR/issue-87493.rs:1:11
+   |
+LL | pub trait MyTrait {
+   |           ^^^^^^^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0107`.

